pennylane/estimator/resource_tracking.py

```diff
@@ -0,0 +1,383 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Core resource tracking logic."""
+from collections import defaultdict
+from collections.abc import Callable, Iterable
+from functools import singledispatch, wraps
+
+from pennylane.operation import Operation
+from pennylane.queuing import AnnotatedQueue, QueuingManager
+from pennylane.wires import Wires
+
+from .resource_config import ResourceConfig
+from .resource_operator import (
+    CompressedResourceOp,
+    GateCount,
+    ResourceOperator,
+)
+from .resources_base import DefaultGateSet, Resources
+from .wires_manager import Allocate, Deallocate, WireResourceManager
+
+# pylint: disable=protected-access,too-many-arguments
+
+
+def map_to_resource_op():  # TODO: Import this function instead when the mapping PR is merged
+    """Maps an instance of :class:`~.Operation` to its associated :class:`~.pennylane.labs.resource_estimation.ResourceOperator`."""
+
+
+def estimate(
+    obj: ResourceOperator | Callable | Resources | list,
+    gate_set: set | None = None,
+    zeroed: int = 0,
+    any_state: int = 0,
+    tight_budget: bool = False,
+    config: ResourceConfig | None = None,
+) -> Resources | Callable:
+    r"""Estimate the quantum resources required from a circuit or operation in terms of the gates
+    provided in the gateset.
+
+    Args:
+        obj (ResourceOperator | Callable | Resources | list): The quantum circuit or operation
+            to obtain resources from.
+        gate_set (set | None): A set of names (strings) of the fundamental operations to track
+            counts for throughout the quantum workflow.
+        zeroed (int | None): Number of zeroed state work wires. Default is ``0``.
+        any_state (int | None): Number of work wires in an unknown state. Default is ``0``.
+        tight_budget (bool | None): Determines whether extra zeroed state wires can be allocated when they
+            exceed the available amount. The default is ``False``.
+        config (ResourceConfig | None): A ResourceConfig object of additional parameters which sets default values
+            when they are not specified on the operator.
```

Comment:
Let's generalize this a bit, no need to be specific here.
```suggestion
        config (ResourceConfig | None): A ResourceConfig object containing details such as
            operator precisions and custom resource decompositions.
```